### PR TITLE
Upgrade to Blitz 2.0 🔥 

### PR DESCRIPTION
**Pull request by nsunami**
This PR upgrades the app to Blitz 2.0 🔥  I started off using the auto-updater ([`npx @blitzjs/codemod upgrade-legacy`](https://blitzjs.com/docs/upgrading-from-framework)), but after that, it was such a ride! Below are the challenges that I encountered, and the solutions taken. Any suggestions are welcome 👍 

# Fixing imports  

There were files that were affected by module not found error (imports). 

One thing to note is that the syntax highlighting theme`a11yLight` from `"react-syntax-highlighter/dist/cjs/styles/hljs` was not available. For now, I switched to `prism` for the light theme. 

# Changing variable name `module` to `currentModule`

ESLint throws an error when a variable is named as `module` since it's a global variable ([`no-implicit-global` rule](https://eslint.org/docs/latest/rules/no-implicit-globals)). I changed them to `currentModule` for this PR.

# Floating promises

ESLint has a rule for [`no-floating-promises`](https://typescript-eslint.io/rules/no-floating-promises/), where we need to specify what happens when there's an error. Our code base had the following floating promises:

- `toast.promise()`
- `router.push()`
- `formik.setFieldValue()`

I changed them to have `.catch(() => {})` to retain our behavior of not returning anything when there's an error. We could improve error handling in the future.

# `Router.push()` no longer working

`Router.push()` was no longer working. I converted to `useRouter()` hook instead. 

# String manipulation

There are some errors about string manipulation as below:

1. `reference.publishedAt.toISOString().substr(0,4)` not working to get the YYYY-MM-DD value. The `publishedAt` here became a string, not Data, and thus `toISOString()` does not work. I fixed so that the code performs `substr()` directly to the string value `reference?.publishedAt?.substr(0, 4)` We get `TypeError: ref.toISOString is not a function` Error, for the published modules with references.
2. `params!.handle?.toLowerCase()` does not work. I fixed this by doing `.toString()` before doing `.toLowerCase()`. 

# Duplicate `changeTitle` mutations

Initially, I could not change the collection title. It turned out that the `changeTitle` mutation was duplicated with the `changeTitle` for modules (and thus crashing namespace). I fixed it by naming the mutation `changeCollectionTitle`

# Validating collection description

The Formik validation for collection description was failing because it was expecting an object that contains a string, but the data was a simple string. I fixed it by making it expect a simple string `validateZodSchema(z.string())`.

# Fonts not loading

At first, font face was not loading property. I had to add font imports in the `Head` element in `_documents.tsx` to fix this.

# Reset password page redirecting, even though when there's a token

The reset password page is set to redirect in an absence of a token. But, initially, the redirect was happening even though with a valid token. This behavior was fixed by using `useRouterQuery`, instead of getting a query string from router (`router.query`). 

# Anti-CSRF token missing 

Our API endpoints for Stripe was not working because of `CSRFTokenMismatchError`. I fixed this by including a hidden form element in the form.

```
<input type="hidden" name="anti-csrf" value={antiCSRFToken} />
```

# App startup is a bit slow

I have not timed this, but I feel like the app start up is a bit slow after the upgrade 🐢 

# Jest-related issues

At first, `npm run test` was failing to run. I had to do the following to fix the tests 

1. Update the `jest.config.js` as per [Blitz documentation](https://blitzjs.com/docs/testing#jest-config)
2. Install additional dependencies, (e.g., `ts-test`, the `@testing-library*` dependencies, `jest`, `jest-environment-jsdom`)

# Build-related issues

In the build time, I had new errors came up. They are:

1. New import errors
2. `ReferenceError: test is not defined` in `index.test.js` - The problem was that the test file was being complied since it was in the `/pages` directory. Next.JS recommends against including tests in the `/pages` folder. Instead, t[hey recommend putting tests in the `__tests__` folder, following Jest convention](https://nextjs.org/docs/testing#:~:text=Your%20project%20is%20now%20ready%20to%20run%20tests.%20Follow%20Jest%27s%20convention%20by%20adding%20tests%20to%20the%20__tests__%20folder%20in%20your%20project%27s%20root%20directory.). I followed this for this file. 
3. After all this, `index.test.js` still fails with a `SyntaxError: Cannot use import statement outside a module`. I was able to replicate this error in the latest Blitz app `blitz new newApp`, and then running `npm run test`. (The files are identical). Since we are not using this test anyway, and this failure may be out of our reach to address, I commented out the code in `index.test.js` for now. 🙏 

Also about the build—I realized these build errors after I checked the review app deployment. Currently, our pre-push hook does not check if the build successes. I wondered if adding `npm run build` to pre-push would be better (or is it too much overhead..?). 💭 

Anyway, this is a long PR! I think I may have missed a bunch of things, so please point out any mistakes or opportunities for improvements. Looking forward to making it better together! 🤝 

----
# Review Checklist 

@chartgerink I'm writing down the points that I want to focus on below for the review! 🙏 

- [x]  Stripe testing
    - [x]  Collections
        - [x]  Collaborative
        - [x]  Community
        - [x]  Upgrades
            - [x]  Personal → Collaborative
            - [x]  Personal → Community
            - [x]  Collaborative → Community
    - [x]  Pay-to-close
- [x]  Markdown code highlighting in the light mode (this can be a design change, so I want you to check it)
- [x]  Connect ORCiD

### CODE REVIEW

**Review by chartgerink — CHANGES_REQUESTED** [review]
We're getting there thanks to your fantastic work on this. It's a lot and I appreciate the effort you have and are putting in :pray: See below for blockers, nice to haves, and remaining questions/comments.

Here's a gif to keep us going:




## Blockers

- [x] verify email not working
```
14:36:43.778 ERROR verify-email() 
 Error  
Invalid `prisma.user.findUnique()` invocation:

{
  where: {
    id: NaN
        ~~~
  },
```
- [x] ORCID authentication didn't work - middleware error - [quick fix tho: wrapping the ORCID api route in `api()` middleware.](https://github.com/blitz-js/blitz/issues/3467#issuecomment-1185676391)
- [ ] Stripe payments not working cuz of Anti-CSRF issue

## Nice to haves

- [ ] Directory tree
	- [ ] Could you provide a short overview of the current directory structure? (also helpful for new contributors!)
	- [ ] Could you remove empty folders?
- [ ] Are there alternatives to the many `.catch(() => {})`? If we could replace those with an `await` or so it would definitely make the code more readably


## Remaining Questions/comments

- [ ] Tested RSS feed?
- [ ] Need to still test Stripe after anti-CSRF is fixed.
- Syntax highlighting looks fine :+1:
- floating promises -> would be good to at least log the error so that we can pinpoint it if something goes wrong. Is there any other way around this as we now repeat something very often that does nothing? (maybe using `await`?)
- You say `Router.push()` is no longer working - any indication why? 
- String manipulation - i reverted the change and did not get an error? It only gets YYYY btw for the references.
- Duplicate `changeTitle` mutations - very good! I think that might have caused some of the issues in the collections too.
- Validating collection description - nice!
- Fonts not loading - removed the loading from the css stylesheet as a result.
- App startup is a bit slow - I notice this too! I figure it'll be better in production as dev tends to be slower. We could think about performance measurements to keep track of this and optimize in the future
- Jest-related issues - we need to add these dependencies to the wiki then!
- build issues - I am confused by the presence of `/__tests__` and `/test` - can we reconcile these?

> I wondered if adding npm run build to pre-push would be better (or is it too much overhead..?). 💭

- I think running build is a lot of overhead indeed. In essence this should not be an issue if everything else works. `npx tsc` should run the tests anyway, and it builds in test deployments too.

File: db/seeds.ts [hunk 1]
```diff
@@ -607,17 +606,17 @@ const seed = async () => {
...
+        name: currentModule.title,
+        description: currentModule.description,
+        publishedAt: currentModule.publishedAt,
       })
     }
   }
 }
 
-export default seed
+export default seed
```
[the remark this answers is not in the document]
Comment (nsunami) [reply db/seeds.ts:622]: This may be my VS Code not doing `files.insertFinalNewline` 😓

**Comment by nsunami** [comment]
Thanks for the review! Below are my progresses:


> ## Blockers


> * [ ]  verify email not working
> 
> ```
> 14:36:43.778 ERROR verify-email() 
>  Error  
> Invalid `prisma.user.findUnique()` invocation:
> 
> {
>   where: {
>     id: NaN
>         ~~~
>   },
> ```

I could not replicate this on my end. I can create an account, and clicking the verification email works fine for me. Could you provide steps to reproduce? 

> * [ ]  ORCID authentication didn't work - middleware error - [quick fix tho: wrapping the ORCID api route in `api()` middleware.](https://github.com/blitz-js/blitz/issues/3467#issuecomment-1185676391)

I pushed https://github.com/libscie/ResearchEquals.com/pull/924/commits/4dc068147a28348abd439929d274c0db22e944c2, and 

> * [ ]  Stripe payments not working cuz of Anti-CSRF issue

The Stripe payments should be working now! 

> ## Nice to haves
> * [ ]  Directory tree
>   
>   * [ ]  Could you provide a short overview of the current directory structure? (also helpful for new contributors!)
>   * [ ]  Could you remove empty folders?

I think Git does not empty folders (and thus I believe empty folders are not pushed to remote--let me know if there are). Would something like doing [git-clean](https://git-scm.com/docs/git-clean/2.22.0) solves this? 

> * [ ]  Are there alternatives to the many `.catch(() => {})`? If we could replace those with an `await` or so it would definitely make the code more readably

Great suggestion! I think just do await makes sense in terms of readability (where catching is not needed) 👍 
 
> ## Remaining Questions/comments
> * [ ]  Tested RSS feed?

Just tested on my localhost and it looks ok 👍 

> * [ ]  Need to still test Stripe after anti-CSRF is fixed.

Now this has been fixed 👍 

> * Syntax highlighting looks fine 👍

> * floating promises -> would be good to at least log the error so that we can pinpoint it if something goes wrong. Is there any other way around this as we now repeat something very often that does nothing? (maybe using `await`?)

Agreed. We can also do `toast.error(e.message)` to provide a more meaningful error message, perhaps together with `console.log(e)`. I tried doing so in our new payment handler: https://github.com/libscie/ResearchEquals.com/pull/924/commits/fcd475582922b7634f41f8c228a2db33f94d1260 

> * You say `Router.push()` is no longer working - any indication why?

`Router` is no longer exported from `blitz`, and thus we cannot use it. Instead, we use NextJS's `useRouter()`. 

> * String manipulation - i reverted the change and did not get an error? It only gets YYYY btw for the references.

Oh, that is weird. Let me check. 

> * Duplicate `changeTitle` mutations - very good! I think that might have caused some of the issues in the collections too.
🙌 

> * Validating collection description - nice!
👍 

> * Fonts not loading - removed the loading from the css stylesheet as a result.

Thanks!

> * App startup is a bit slow - I notice this too! I figure it'll be better in production as dev tends to be slower. We could think about performance measurements to keep track of this and optimize in the future

Yup!

> * Jest-related issues - we need to add these dependencies to the wiki then!

Got it!

> * build issues - I am confused by the presence of `/__tests__` and `/test` - can we reconcile these?

Probably, the best way to move all tests to `/__tests__` following the Jest's convention

> 
> > I wondered if adding npm run build to pre-push would be better (or is it too much overhead..?). 💭
> 
> * I think running build is a lot of overhead indeed. In essence this should not be an issue if everything else works. `npx tsc` should run the tests anyway, and it builds in test deployments too.

Got it!


# Nami's To-Do's 

- [ ] Overview of the directory structure 
- [ ] Converting the floating promises to await
- [ ] Check the string manipulation issues (YYYY?)
- [ ] Add Jest-related dependencies to Wiki
- [ ] Move test files in `/test` to `__tests__`

**Review by chartgerink — CHANGES_REQUESTED** [review]
This is looking great! 🙏 It is mergeable, save for one thing that needs to be fixed. 1️⃣ 

**almost there!!!!**

## Dashboard does not redirect to log in screen

Right now when I go to the dashboard (when logged out) I get this prisma error:
Screenshot 2022-11-28 at 17 10 59


In production, I do get the regular login window when I go to the https://researchequals.com/dashboard page while logged out.
Screenshot 2022-11-28 at 17 11 22


It's fine for e.g., https://researchequals.com/admin in both production and testing, so I'm not sure where this is coming from 🤔

**Comment by nsunami** [comment]
Thanks, Chris! 👍 

I tried solving the error on the `/dashboard` when logged out, as follows:

1. https://github.com/libscie/ResearchEquals.com/pull/924/commits/db5a44c831192ae3bbf6a7c762b3ca6dd1c98806 We are now passing `undefined` as `userId` when logged out, and passing this to the DB query. 

2. https://github.com/libscie/ResearchEquals.com/pull/924/commits/962a04c86ed860368504df54b1230db4089736ae After testing this change on GitHub Codespaces, I got another 500 error, `(intermediate value) is not iterable.` in `getDashboardData`. The cause of this issue was that we were trying to ignore the null in this line `...workspace?.following.map((x) => x.id)!` and we did not have a fallback when it was null. To solve this, I added an empty array as a fallback: `...(workspace?.following.map((x) => x.id) ?? [])`

(*Side note: The commit https://github.com/libscie/ResearchEquals.com/pull/924/commits/962a04c86ed860368504df54b1230db4089736ae was not signed because I did not have a GPG key setup on my Codespaces env. 😓 )

Hope this works!

**Review by chartgerink — APPROVED** [review]

**Pull request merged**